The process with PID could not be terminated. The operation attempted is not supported.

Sometimes the ES instance on Windows can't be killed, and the build fails :-(

May be related to an [unkillable state](https://blogs.technet.microsoft.com/markrussinovich/2005/08/17/unkillable-processes/).

```
[INFO] com.github.alexcojocaru.mojo.elasticsearch.v2.ElasticsearchSetupException: Elasticsearch [0]: Cannot execute command '[taskkill, /F, /pid, 12340]' in directory 'C:\Projects\elasticsearch-maven-plugin\target\it\runforked-with-path-conf\target\elasticsearch0'
[INFO] Output:
[INFO]
[INFO] Error:
[INFO] ERROR: The process with PID 12340 could not be terminated.
[INFO] Reason: The operation attempted is not supported.
[INFO]
[INFO]  at com.github.alexcojocaru.mojo.elasticsearch.v2.util.ProcessUtil.executeScript(ProcessUtil.java:267)
[INFO]  at com.github.alexcojocaru.mojo.elasticsearch.v2.util.ProcessUtil.executeScript(ProcessUtil.java:149)
[INFO]  at com.github.alexcojocaru.mojo.elasticsearch.v2.ForkedElasticsearchProcessDestroyer.terminateWindowsProcess(ForkedElasticsearchProcessDestroyer.java:118)
[INFO]  at com.github.alexcojocaru.mojo.elasticsearch.v2.ForkedElasticsearchProcessDestroyer.terminateProcess(ForkedElasticsearchProcessDestroyer.java:65)
[INFO]  at com.github.alexcojocaru.mojo.elasticsearch.v2.ForkedElasticsearchProcessDestroyer.access$000(ForkedElasticsearchProcessDestroyer.java:15)
[INFO]  at com.github.alexcojocaru.mojo.elasticsearch.v2.ForkedElasticsearchProcessDestroyer$1.run(ForkedElasticsearchProcessDestroyer.java:28)
[INFO] Caused by: org.apache.commons.exec.ExecuteException: Process exited with an error: 1 (Exit value: 1)
[INFO]  at org.apache.commons.exec.DefaultExecutor.executeInternal(DefaultExecutor.java:404)
[INFO]  at org.apache.commons.exec.DefaultExecutor.execute(DefaultExecutor.java:166)
[INFO]  at com.github.alexcojocaru.mojo.elasticsearch.v2.util.ProcessUtil.executeScript(ProcessUtil.java:222)
[INFO]  ... 5 more
```

--- a/src/main/java/com/github/alexcojocaru/mojo/elasticsearch/v2/ForkedElasticsearchProcessDestroyer.java
+++ b/src/main/java/com/github/alexcojocaru/mojo/elasticsearch/v2/ForkedElasticsearchProcessDestroyer.java
@@ -113,29 +113,45 @@ public class ForkedElasticsearchProcessDestroyer implements ProcessDestroyer
                 config.getId()));
         
         CommandLine command = ProcessUtil.buildKillCommandLine(pid);
-        try
-        {
-            ProcessUtil.executeScript(config, command, true);
-
-            log.info(String.format(
-                    "... the Elasticsearch process [%d] has stopped.",
-                    config.getId()));
-        }
-        catch (Exception e)
-        {
-            if (e.getMessage().contains("no running instance"))
+        for (int retry = 0; retry < 3; ++retry) {
+            try
             {
+                ProcessUtil.executeScript(config, command, true);
+    
                 log.info(String.format(
                         "... the Elasticsearch process [%d] has stopped.",
                         config.getId()));
+                ProcessUtil.cleanupPid(config.getBaseDir());
             }
-            else
+            catch (Exception e)
             {
-                log.error(String.format(
-                        "Error while destroying the Elasticsearch process [%d]",
-                        config.getId()),
-                        e);
+                if (e.getMessage().contains("no running instance"))
+                {
+                    log.info(String.format(
+                            "... the Elasticsearch process [%d] has stopped.",
+                            config.getId()));
+                    ProcessUtil.cleanupPid(config.getBaseDir());
+                }
+                else
+                {
+                    // sometimes taskkill fails with "The operation attempted is not supported."
+                    // https://blogs.technet.microsoft.com/markrussinovich/2005/08/17/unkillable-processes/
+                    // in that case, retry
+                    if (retry < 2) {
+                        try {
+                            Thread.sleep(500);
+                            continue;
+                        } catch (InterruptedException e1) {
+                            /* ignore */
+                        }
+                    }
+                    log.error(String.format(
+                            "Error while destroying the Elasticsearch process [%d]",
+                            config.getId()),
+                            e);
+                }
             }
+            break;
         }
     }
     

--- a/src/main/java/com/github/alexcojocaru/mojo/elasticsearch/v2/ForkedInstance.java
+++ b/src/main/java/com/github/alexcojocaru/mojo/elasticsearch/v2/ForkedInstance.java
@@ -39,7 +39,7 @@ public class ForkedInstance
 
         ProcessUtil.executeScript(config,
                 getStartScriptCommand(),
-                null,
+                ProcessUtil.createEnvironment(null),
                 new ForkedElasticsearchProcessDestroyer(config));
     }
 

--- a/src/main/java/com/github/alexcojocaru/mojo/elasticsearch/v2/StopMojo.java
+++ b/src/main/java/com/github/alexcojocaru/mojo/elasticsearch/v2/StopMojo.java
@@ -39,6 +39,7 @@ public class StopMojo
                 ProcessUtil.executeScript(config, getShutdownScriptCommand(baseDir));
 
                 getLog().info(String.format("Elasticsearch [%d] stopped", config.getId()));
+                ProcessUtil.cleanupPid(baseDir);
             }
             catch (Exception e)
             {

--- a/src/main/java/com/github/alexcojocaru/mojo/elasticsearch/v2/step/InstallPluginsStep.java
+++ b/src/main/java/com/github/alexcojocaru/mojo/elasticsearch/v2/step/InstallPluginsStep.java
@@ -54,7 +54,7 @@ public class InstallPluginsStep
                     .addArgument("--batch")
                     .addArgument(plugin.getUri());
             
-            ProcessUtil.executeScript(config, cmd, environment, null);
+            ProcessUtil.executeScript(config, cmd, ProcessUtil.createEnvironment(environment), null);
         }
     }
 }

--- a/src/main/java/com/github/alexcojocaru/mojo/elasticsearch/v2/util/ProcessUtil.java
+++ b/src/main/java/com/github/alexcojocaru/mojo/elasticsearch/v2/util/ProcessUtil.java
@@ -76,6 +76,18 @@ public class ProcessUtil
         
         return command;
     }
+    
+    /**
+     * After forcibly killing the ES instance, remove the pid file.
+     * @param baseDir the base directory where the pid file is
+     */
+    public static void cleanupPid(String baseDir)
+    {
+        if (SystemUtils.IS_OS_WINDOWS)
+        {
+            new File(baseDir, "pid").delete();
+        }        
+    }
 
     /**
      * Read the ES PID from the pid file and return it.
@@ -293,7 +305,7 @@ public class ProcessUtil
      * @param environment
      * @return an execution environment
      */
-    private static Map<String, String> createEnvironment(Map<String, String> environment)
+    public static Map<String, String> createEnvironment(Map<String, String> environment)
     {
         Map<String, String> result = null;
         
@@ -316,6 +328,11 @@ public class ProcessUtil
         // and unsets these. And because the scripts would print a warning, we can't rely on the output :(
         result.remove("JAVA_TOOL_OPTIONS");
         result.remove("JAVA_OPTS");
+        // the following environment variables may interfere with the elasticsearch instance.
+        result.remove("ES_HOME");
+        result.remove("ES_JAVA_OPTS");
+        result.remove("ES_PATH_CONF");
+        result.remove("ES_TMPDIR");
 
         return result;
     }

--- a/src/test/java/com/github/alexcojocaru/mojo/elasticsearch/v2/step/BootstrapClusterStepTest.java
+++ b/src/test/java/com/github/alexcojocaru/mojo/elasticsearch/v2/step/BootstrapClusterStepTest.java
@@ -101,7 +101,7 @@ public class BootstrapClusterStepTest
         
         ArgumentCaptor<Path> pathCaptor = ArgumentCaptor.forClass(Path.class);
         verify(step).parseJson(any(ElasticsearchClient.class), eq(log), pathCaptor.capture());
-        assertEquals(filePath, pathCaptor.getValue().toString());
+        assertEquals(filePath, pathCaptor.getValue().toString().replace('\\', '/'));
     }
 
     @Test
@@ -121,7 +121,7 @@ public class BootstrapClusterStepTest
         
         ArgumentCaptor<Path> pathCaptor = ArgumentCaptor.forClass(Path.class);
         verify(step).parseScript(any(ElasticsearchClient.class), eq(log), pathCaptor.capture());
-        assertEquals(filePath, pathCaptor.getValue().toString());
+        assertEquals(filePath, pathCaptor.getValue().toString().replace('\\', '/'));
     }
     
     @Test(expected = ElasticsearchSetupException.class)

